File: CreationalPatterns/Prototype/Java/Interfaces/AbstractPizza.java
package Interfaces;

import java.util.ArrayList;
import java.util.List;

import Prototype.IPizzaPrototype;

public abstract class AbstractPizza implements IPizzaPrototype {
    public String dough;
    public String sauce;
    public List<String> toppings;

    public AbstractPizza(){
        this.toppings = new ArrayList<>();
    }

    protected AbstractPizza(AbstractPizza source){
        this.dough = source.dough;
        this.sauce = source.sauce;
        this.toppings = new ArrayList<>(source.toppings);
    }

    public void setDough(String dough) { this.dough = dough; }
    public void setSauce(String sauce) { this.sauce = sauce; }
    public void addTopping(String topping) { this.toppings.add(topping); }
    
    @Override
    public String toString() {
        return "Pizza with " + dough + " dough, " + sauce + " sauce, and toppings: " + toppings;
    }
}

File: CreationalPatterns/Prototype/C#/Prototype/Interfaces/AbstractPizza.cs
public abstract class AbstractPizza : IPizzaPrototype
{
    public String? Dough { get; set; }
    public String? Sauce { get; set; }
    public List<String> Toppings { get; set; }

    public AbstractPizza()
    {
        this.Toppings = new List<string>();
    }

    protected AbstractPizza(AbstractPizza source)
    {
        this.Dough = source.Dough;
        this.Sauce = source.Sauce;
        this.Toppings = new List<string>(source.Toppings);
    }

    public void SetDough(String dough) { this.Dough = dough; }
    public void SetSauce(String sauce) { this.Sauce = sauce; }
    public void AddTopping(String topping) { this.Toppings.Add(topping); }

    public override String ToString()
    {
        return "Pizza with " + Dough + " dough, " + Sauce + " sauce, and toppings: " + Toppings;
    }
    
    public abstract IPizzaPrototype Copy();
}
